src/main.cpp

```diff
@@ -18,104 +18,183 @@
+
+	// Show the user the valid device IDs they can use.
+	auto device_list = SdlAudioSink::GetDevicesInfo();
+	for (const auto &device : device_list) {
+		std::cerr << "\t"
+			  << device.first << ": " << device.second
+			  << "\n";
 	}
-	this->audio.SetDeviceID(id);
 
-	// Now set up all the IO (network socket and event loop).
-	try {
-		this->io = decltype(this->io)(new IoCore(
-		                this->player, this->handler, addr, port));
-	} catch (NetError &e) {
-		std::cerr << "Network error: " << e.Message() << std::endl;
-		std::cerr << "Is " << addr << ":" << port << " available?"
-		          << std::endl;
-		return EXIT_FAILURE;
-	}
+	std::cerr << "default HOST: " << DEFAULT_HOST << "\n";
+	std::cerr << "default PORT: " << DEFAULT_PORT << "\n";
 
-	try {
-		this->io->Run();
-	} catch (Error &error) {
-		std::cerr << "Unhandled exception in main loop: "
-		          << error.Message() << std::endl;
-		return EXIT_FAILURE;
-	}
+	exit(EXIT_FAILURE);
+}
 
-	return EXIT_SUCCESS;
+/**
+ * Gets the host and port from the program arguments.
+ * The default arguments are used if the host and/or port are not supplied.
+ * @param args The program argument vector.
+ * @return A pair of strings representing the hostname and port.
+ */
+std::pair<std::string, std::string> GetHostAndPort(const std::vector<std::string> &args) {
+	auto size = args.size();
+	return std::make_pair(
+		size > 2 ? args.at(2) : DEFAULT_HOST,
+		size > 3 ? args.at(3) : DEFAULT_PORT
+	);
+}
+
+/**
+ * Exits with an error message for a network error.
+ * @param host The IP host to which playd tried to bind.
+ * @param port The TCP port to which playd tried to bind.
+ * @param msg The exception's error message.
+ */
+[[noreturn]] void ExitWithNetError(const std::string &host, const std::string &port, const std::string &msg) {
+	std::cerr << "Network error: " << msg << "\n";
+	std::cerr << "Is " << host << ":" << port << " available?\n";
+	exit(EXIT_FAILURE);
 }
 
-void Playd::RespondRaw(const std::string &string) const
+/**
+ * Exits with an error message for an unhandled exception.
+ * @param msg The exception's error message.
+ */
+[[noreturn]] void ExitWithError(const std::string &msg) {
+	std::cerr << "Unhandled exception in main loop: "
+		  << msg << std::endl;
+	exit(EXIT_FAILURE);
+}
+
+/**
+ * The main entry point.
+ * @param argc Program argument count.
+ * @param argv Program argument vector.
+ * @return The exit code (zero for success; non-zero otherwise).
+ */
+int main(int argc, char *argv[])
 {
-	if (this->io != nullptr) this->io->Broadcast(string);
+	// This call needs to happen before GetDeviceID, otherwise no device
+	// IDs will be recognised.  (This is why it's here, and not in
+	// SetupAudioSystem.)
+	SdlAudioSink::InitLibrary();
+	atexit(SdlAudioSink::CleanupLibrary);
+
+	auto args = MakeArgVector(argc, argv);
+
+	auto device_id = GetDeviceID(args);
+	if (device_id < 0) ExitWithUsage(args.at(0));
```

Comment:
now, these functions.
Unnecessary, i think. They're only used once, are largely duplicated, and functions can easily be reordered to avoid the atexit stuff, i think